### PR TITLE
Enable mmap-ped Faiss indices (IndexFlatCodes-based and HNSW)

### PR DESCRIPTION
**Pull request by alexanderguzhva**

### CODE REVIEW

File: thirdparty/faiss/faiss/impl/mapped_io.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        void* address = mmap(nullptr, filesize, PROT_READ, MAP_SHARED, fileno(f), 0);
+        FAISS_THROW_IF_NOT_FMT(
+            address != nullptr, "could not mmap(): %s", strerror(errno)
+        );
+
+        // btw, fd can be closed here
+
+        // set 'random' access pattern
+        // todo: check the error
+        madvise(address, filesize, MADV_RANDOM);
```
Comment (foxspy) [inline thirdparty/faiss/faiss/impl/mapped_io.cpp:43]: a warn log needs to be added here

File: thirdparty/faiss/faiss/impl/index_read.cpp [hunk 2]
```diff
@@ -512,7 +594,8 @@ static void read_HNSW(HNSW* hnsw, IOReader* f) {
     READVECTOR(hnsw->cum_nneighbor_per_level);
     READVECTOR(hnsw->levels);
```
Comment (foxspy) [inline thirdparty/faiss/faiss/impl/index_read.cpp:595]: Don't these require mmap? Although these only account for a small part

**Review by foxspy — APPROVED** [review]
/lgtm

File: thirdparty/faiss/faiss/impl/index_read.cpp (hunk 2 above)
Comment (chasingegg) [reply thirdparty/faiss/faiss/impl/index_read.cpp:595]: same question, what do we exactly mmap for hnsw @alexanderguzhva

**Pull request merged**